c++/test/TestPredicatePushdown.cc

```diff
@@ -0,0 +1,183 @@
+    {
+      std::unique_ptr<SearchArgument> sarg = SearchArgumentFactory::newBuilder()
+        ->startAnd()
+        .startNot()
+        .lessThan("x", PredicateDataType::LONG,
+                  Literal(static_cast<int64_t>(300000L)))
+        .end()
+        .lessThan("x", PredicateDataType::LONG,
+                  Literal(static_cast<int64_t>(600000L)))
+        .end()
+        .build();
+
+      RowReaderOptions rowReaderOpts;
+      rowReaderOpts.setSearchArgument(std::move(sarg));
+      auto rowReader = reader->createRowReader(rowReaderOpts);
+
+      auto readBatch = rowReader->createRowBatch(2000);
+      auto& batch0 = dynamic_cast<StructVectorBatch&>(*readBatch);
+      auto& batch1 = dynamic_cast<LongVectorBatch&>(*batch0.fields[0]);
+      auto& batch2 = dynamic_cast<StringVectorBatch&>(*batch0.fields[1]);
+
+      EXPECT_EQ(true, rowReader->next(*readBatch));
+      EXPECT_EQ(1000, readBatch->numElements);
+      EXPECT_EQ(1000, rowReader->getRowNumber());
+      for (uint64_t i = 1000; i < 2000; ++i) {
+        EXPECT_EQ(300 * i, batch1.data[i - 1000]);
+        EXPECT_EQ(std::to_string(10 * i),
+          std::string(batch2.data[i - 1000], static_cast<size_t>(batch2.length[i - 1000])));
+      }
+      EXPECT_EQ(false, rowReader->next(*readBatch));
+      EXPECT_EQ(3500, rowReader->getRowNumber());
+    }
+
+    // look through the file with no rows selected: x < 0
+    {
+      std::unique_ptr<SearchArgument> sarg = SearchArgumentFactory::newBuilder()
+        ->startAnd()
+        .lessThan("x", PredicateDataType::LONG,
+          Literal(static_cast<int64_t>(0)))
+        .end()
+        .build();
+
+      RowReaderOptions rowReaderOpts;
+      rowReaderOpts.setSearchArgument(std::move(sarg));
+      auto rowReader = reader->createRowReader(rowReaderOpts);
+
+      auto readBatch = rowReader->createRowBatch(2000);
+      EXPECT_EQ(false, rowReader->next(*readBatch));
+      EXPECT_EQ(3500, rowReader->getRowNumber());
+    }
+
+    // select first 100 and last 100 rows: x < 30000 OR x >= 1020000
+    {
+      std::unique_ptr<SearchArgument> sarg = SearchArgumentFactory::newBuilder()
+        ->startOr()
+        .lessThan("x", PredicateDataType::LONG,
+          Literal(static_cast<int64_t>(300 * 100)))
+        .startNot()
+        .lessThan("x", PredicateDataType::LONG,
+          Literal(static_cast<int64_t>(300 * 3400)))
+        .end()
+        .end()
+        .build();
+
+      RowReaderOptions rowReaderOpts;
+      rowReaderOpts.setSearchArgument(std::move(sarg));
+      auto rowReader = reader->createRowReader(rowReaderOpts);
+
+      auto readBatch = rowReader->createRowBatch(2000);
+      auto& batch0 = dynamic_cast<StructVectorBatch&>(*readBatch);
+      auto& batch1 = dynamic_cast<LongVectorBatch&>(*batch0.fields[0]);
+      auto& batch2 = dynamic_cast<StringVectorBatch&>(*batch0.fields[1]);
+
+      EXPECT_EQ(true, rowReader->next(*readBatch));
+      EXPECT_EQ(1000, readBatch->numElements);
```

Comment:
I would assume 2 row-groups selected based on the comment above

Reply:
The above comment for pred is slightly wrong. But still 2 row groups are selected.